### ISSUE TITLE
chase down API access to ACRA BizFile/BizHub

### ISSUE THREAD

**Issue by mengwong** [legalese-io/legalese-io.github.io#119]
> Category:General Enquiries (other than XBRL & UEN)
> Question:
> 
> > Hi, in reference to https://www.acra.gov.sg/components/wireframes/howToGuidesChapters.aspx?pageid=1262#1274 I am writing on behalf of a corporate service provider. I would like to run automated queries against the ACRA Web Services Gateway. For example, when a user comes to my website and types in their company name or UEN, I want my website to be able to run a "Retrieval of Basic Company’s Profile" against ACRA using the gateway. Where can I find more information about the API, and how can I establish an account with the ACRA Web Services Gateway to connect my website with your website? According to the information on that page I may have to go through GWS-X. I have read http://www.egov.gov.sg/egov-masterplans/egap-ii/programmes/government-web-services-exchange and it looks promising. However I saw a note: "With effect from January 2014, subscription services for the services below have been temporarily suspended as we are currently revamping our business information services system. Please visit this page for future updates." so I'm not sure what to think now. Thanks in advance for your assistance, happy holidays.
> 
> Thank you for your enquiry. We will look into the matter promptly and get back to you within 3 working days. However, depending on the nature and complexity of the enquiries, we may need up to 14 working days to review some cases.
> 
> Please take note of the reference number assigned to your enquiry:
> AA1512-041264

later:

> Date: Thursday, 31 December 2015
> Dear Mr. Wong Meng Weng,
> We thank you for your interest to subscribe to our Web Services.
> Unfortunately, due to ongoing activities on our Enhanced Bizfile+ project (due to launch on 3 Jan 2016), our subscription services request is currently being suspended until further notice.
> For more information, please visit our ACRA portal.

Intended use:

The end-user's web browser will talk to our web server. Our web server will be a client to ACRA's web server. Our web server will ask ACRA to turn a company name into a list of UENs and addresses. To do that, we will have to include our credentials in the request.

We will get a response back from ACRA; we will reformat the response and hand it back to the end-user's web browser.

Subsequently, the end-user will click on something, which will cause our web server to then ask ACRA to turn a UEN into a list of shareholders. When we get the list back, we will reformat the response, possibly by sticking it into Google Spreadsheets, and hand some gesture of progress back to the end-user's web browser.

So, that's the high level goal of our moving parts.

what we are looking for is documentation that looks like https://secure.na1.echosign.com/public/docs/restapi/v5

or documentation like http://api.corpwatch.org/documentation/api_examples.html

or https://developer.companieshouse.gov.uk/api/docs/index/gettingStarted.html
particularly that last one. that is the sort of stick which they are amenable to being beaten with.
https://developer.companieshouse.gov.uk/api/docs/company/company_number/companyProfile-resource.html

In Estonia,
http://www.rik.ee/en/e-business-registry/xml-service

**Comment by wabisabiyo** [comment]
tl;dr:  need to wait for tech team to call me back

**call made to ACRA today.**

**_q:  how do i get more information about ACRA web services or the documentation to implement the API?_**

**ACRA:**  you have to subscribe first before you can get information about that
**Alexis:** that doesn't make sense. i don't want to pay money first. what if it's not what i want? is it not published anywhere that i can access?
**ACRA:** we (phone operators) are unable to provide further details on documentation or anything to do with ACRA web services because we are not tasked with answering such questions. AND we have no internet access.
**Alexis:** oh dear. so what do i do? i don't want to pay money for the elite subscription, unless i know what it entails and it can help me do what i want to do
**ACRA**: how about i request it for you? i can ask the tech team. and they will call you back. they are the ones that provide subscribers with documentation and information like that
**Alexis**: Oooh. Sgtm. I'll wait for your call.

**Comment by mengwong** [comment]
https://www.techinasia.com/singapore-government-api will take some time … maybe the data api and the transactional api will be ready around 2017, 2018 ish. the singpass oauth idea will be up to IDA, but we need to think through the user experience and then discuss the back-end moving parts with IDA and ACRA; how do we present a simplified yet secure yet multi-party experience?

**Comment by mengwong** [comment]
https://www.questnet.sg/ must have a back-end channel to ACRA, because it supports a query interface offering the same data. It also has no captchas, no user-hostile javascript, and no delay.

Next action: write a wrapper library.

**Comment by yosriady** [comment]
@mengwong I have a similar use case, did you guys manage to get API access to ACRA business profile information?

**Comment by mengwong** [comment]
We had a meeting at ACRA last year. As I understand it, ACRA is not the only government body that wants to offer an API. So all the government bodies are going to bring up APIs together, in a unified, consistent way. Such a unified API is at least still a year away. We asked if there was some simpler API available that we can call. They looked at one another, paused, and told us that no such thing existed.

Shortly after the meeting, we discovered that https://www.questnet.sg/ offers an interface which appears to call the simpler API which does not officially exist. We also discovered that the Questnet API offers content in a highly structured HTML format. So we built a little library to interface with that. @jobchong may be able to indicate where that library may be found.

**Comment by mengwong** [comment]
As a point of contrast, note how transparent and open other countries are, e.g. http://coys.co.nz/company/?no=3250067-POTATO+POTATO+FACTORY+LIMITED

**Comment by jobchong** [comment]
Hi @yosriady, there is a CLI tool for this at https://github.com/legalese/legalese-google-app/tree/master/questnet

**Comment by bjornlll** [comment]
Is there any other way of accessing ACRA's APIs through UEN queries?

**Comment by jobchong** [comment]
Hey @bjornlll, there isn't. ACRA's API for business profile information is currently only accessible by a few third-party providers, www.questnet.sg among them. Legalese does not have direct access to the API.

An alternative is a data.gov.sg API for a smaller set of company information, accessible at https://data.gov.sg/api/action/datastore_search?q={%22entity_name%22:%20%22companyyourelookingfor%22}&resource_id=bdb377b8-096f-4f86-9c4f-85bad80ef93c

Hope that this answers your question!

**Comment by bjornlll** [comment]
Thanks @jobchong, that's really helpful.

**Comment by Knowyour** [comment]
Hello, 
Trying to have access to ACRA's API for the business profile (Company Information, Directors, Shareholders ....), it was forecasted for 2017, right? I'm looking how to avoid all CAPTCHA work and have access to the information. Any suggestion?

Thank you

**Comment by cgeissendoerfer** [comment]
I just discovered that an ACRA API Mall was added to the list of CorpPass services (https://www.corppass.gov.sg/corppass/manageeservices/entity/selecteservices).
![screen shot 2018-07-27 at 4 09 52 pm](https://user-images.githubusercontent.com/2594859/43312462-8e66f53c-91b7-11e8-8224-dd08c231f343.png)
No further explanation was added. @mengwong - have you seen this?

**Comment by bjornlll** [comment]
That's super interesting. No other information as to what it would be about? I really did not expect them to roll out any form of API (even thought they did promise they'd do so).

**Comment by mengwong** [comment]
ACRA's API Mall appears to now be in closed beta. Until it opens up more widely, data.gov.sg offers a couple of ACRA-originated datasets, such as https://data.gov.sg/dataset/acra-information-on-corporate-entities

To get access to the closed beta, try connecting on Twitter. https://twitter.com/acra_sg/status/739766012661223424 shows there is an @ACRA_SG account.
